File: frontend/src/lib/workbench.ts
import { API } from "@/lib/api";
import { authenticatedFetch } from "@/lib/authenticated-fetch";

export type RequirementStatus = "draft" | "active" | "planned" | "completed" | "archived";
export type Priority = "low" | "medium" | "high" | "critical";
export type SolutionStatus = "draft" | "active" | "deprecated" | "archived";

export interface Requirement {
  id: number;
  title: string;
  description: string;
  status: RequirementStatus;
  priority: Priority;
  source_type: string;
  source_id: string | null;
  source_url: string | null;
  solution_count?: number;
  solutions?: Solution[];
  created_at: string;
  updated_at: string;
}

export interface Solution {
  id: number;
  name: string;
  description: string;
  category: string;
  status: SolutionStatus;
  version: string;
  reference_url: string | null;
  requirement_count?: number;
  requirements?: Requirement[];
  created_at: string;
  updated_at: string;
}

export const requirementStatusLabels: Record<RequirementStatus, string> = {
  draft: "草稿", active: "进行中", planned: "已规划", completed: "已完成", archived: "已归档",
};
export const priorityLabels: Record<Priority, string> = {
  low: "低", medium: "中", high: "高", critical: "紧急",
};
export const solutionStatusLabels: Record<SolutionStatus, string> = {
  draft: "草稿", active: "使用中", deprecated: "已弃用", archived: "已归档",
};

export async function workbenchFetch<T>(path: string, init?: RequestInit): Promise<T> {
  const response = await authenticatedFetch(`${API}/api/workbench${path}`, init);
  if (!response.ok) {
    const data = await response.json().catch(() => ({}));
    throw new Error(data.detail || `请求失败：${response.status}`);
  }
  if (response.status === 204) return undefined as T;
  return response.json();
}

export const formatWorkbenchDate = (value: string) => new Date(value).toLocaleString("zh-CN", {
  month: "2-digit", day: "2-digit", hour: "2-digit", minute: "2-digit",
});

File: backend/routers/workbench.py
"""Authenticated Requirements and Solutions workbench API."""
from __future__ import annotations

import asyncio
from typing import Literal, Optional, Set

from fastapi import APIRouter, Depends, Response, status
from pydantic import BaseModel, Field, HttpUrl

from routers.auth import require_auth
from services import workbench_service as service

router = APIRouter(prefix="/workbench")
RequirementStatus = Literal["draft", "active", "planned", "completed", "archived"]
Priority = Literal["low", "medium", "high", "critical"]
SolutionStatus = Literal["draft", "active", "deprecated", "archived"]


class RequirementCreate(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    description: str = Field(default="", max_length=5000)
    status: RequirementStatus = "draft"
    priority: Priority = "medium"
    source_type: str = Field(default="manual", max_length=50)
    source_id: Optional[str] = Field(default=None, max_length=200)
    source_url: Optional[HttpUrl] = None


class RequirementUpdate(BaseModel):
    title: Optional[str] = Field(default=None, min_length=1, max_length=200)
    description: Optional[str] = Field(default=None, max_length=5000)
    status: Optional[RequirementStatus] = None
    priority: Optional[Priority] = None
    source_type: Optional[str] = Field(default=None, max_length=50)
    source_id: Optional[str] = Field(default=None, max_length=200)
    source_url: Optional[HttpUrl] = None


class SolutionCreate(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    description: str = Field(default="", max_length=5000)
    category: str = Field(default="未分类", max_length=100)
    status: SolutionStatus = "draft"
    version: str = Field(default="v0.1.0", max_length=50)
    reference_url: Optional[HttpUrl] = None
    requirement_id: Optional[int] = Field(default=None, gt=0)


class SolutionUpdate(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=200)
    description: Optional[str] = Field(default=None, max_length=5000)
    category: Optional[str] = Field(default=None, max_length=100)
    status: Optional[SolutionStatus] = None
    version: Optional[str] = Field(default=None, max_length=50)
    reference_url: Optional[HttpUrl] = None


def _data(model, *, exclude: Optional[Set[str]] = None, unset: bool = False) -> dict:
    data = model.model_dump(exclude=exclude or set(), exclude_unset=unset)
    return {key: str(value) if isinstance(value, HttpUrl) else value for key, value in data.items()}


@router.get("/summary")
async def summary(user=Depends(require_auth)):
    return await asyncio.to_thread(service.get_summary, str(user.id))


@router.get("/requirements")
async def requirements(status: Optional[RequirementStatus] = None, user=Depends(require_auth)):
    return await asyncio.to_thread(service.list_requirements, str(user.id), status)


@router.post("/requirements", status_code=status.HTTP_201_CREATED)
async def create_requirement(payload: RequirementCreate, user=Depends(require_auth)):
    return await asyncio.to_thread(service.create_requirement, str(user.id), _data(payload))


@router.get("/requirements/{requirement_id}")
async def requirement(requirement_id: int, user=Depends(require_auth)):
    return await asyncio.to_thread(service.get_requirement, str(user.id), requirement_id)


@router.put("/requirements/{requirement_id}")
async def update_requirement(requirement_id: int, payload: RequirementUpdate, user=Depends(require_auth)):
    return await asyncio.to_thread(service.update_requirement, str(user.id), requirement_id, _data(payload, unset=True))


@router.delete("/requirements/{requirement_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_requirement(requirement_id: int, user=Depends(require_auth)):
    await asyncio.to_thread(service.delete_requirement, str(user.id), requirement_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.post("/requirements/{requirement_id}/solutions/{solution_id}", status_code=status.HTTP_204_NO_CONTENT)
async def link_solution(requirement_id: int, solution_id: int, user=Depends(require_auth)):
    await asyncio.to_thread(service.link_solution, str(user.id), requirement_id, solution_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.delete("/requirements/{requirement_id}/solutions/{solution_id}", status_code=status.HTTP_204_NO_CONTENT)
async def unlink_solution(requirement_id: int, solution_id: int, user=Depends(require_auth)):
    await asyncio.to_thread(service.unlink_solution, str(user.id), requirement_id, solution_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.get("/solutions")
async def solutions(user=Depends(require_auth)):
    return await asyncio.to_thread(service.list_solutions, str(user.id))


@router.post("/solutions", status_code=status.HTTP_201_CREATED)
async def create_solution(payload: SolutionCreate, user=Depends(require_auth)):
    data = _data(payload, exclude={"requirement_id"})
    return await asyncio.to_thread(service.create_solution, str(user.id), data, payload.requirement_id)


@router.get("/solutions/{solution_id}")
async def solution(solution_id: int, user=Depends(require_auth)):
    return await asyncio.to_thread(service.get_solution, str(user.id), solution_id)


@router.put("/solutions/{solution_id}")
async def update_solution(solution_id: int, payload: SolutionUpdate, user=Depends(require_auth)):
    return await asyncio.to_thread(service.update_solution, str(user.id), solution_id, _data(payload, unset=True))


@router.delete("/solutions/{solution_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_solution(solution_id: int, user=Depends(require_auth)):
    await asyncio.to_thread(service.delete_solution, str(user.id), solution_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)
